added session task to create db table

dberry37388 opened the pull request:
This is a simple oil task that creates, deletes or truncates the sessions table when using db as the driver.

Commit:
Oil task to create the session db table.

Commit:
updated closing comment. added new line at end of file.

philsturgeon commented:
Why?

dberry37388 commented:
Well, Oil is meant to automate tasks, this automates the task of setting up the session database table.  Is it borderline laziness? yes, but then again that is what Oil is about. but I was in the IRC channel and after several times hearing people ask about it decided to throw it together.  90% of the people using FuelPHP are probably using sessions at some point, and 99% of those people would rather run a task than have to go to the docs to copy/paste the table structure, especially if they are already sitting right there in front of the cmd line.

-----Original Message-----
From: Phil Sturgeon [mailto:reply@reply.github.com] 
Sent: Monday, January 23, 2012 3:14 PM
To: dberry37388
Subject: Re: [core] added session task to create db table (#786)

Why?

---

Reply to this email directly or view it on GitHub:
https://github.com/fuel/core/pull/786#issuecomment-3622002

philsturgeon commented:
Ok the up/down threw me a little there and made the whole thing seem a little off.

Tell you what, make it "add/remove" and maybe throw in "clear" to empty the sessions table, and we've got a deal! :)

Commit:
Changed up/down/truncate to create/remove/clear.  Also added a run() method that will display a menu if the user fails to enter an argument.

dberry37388 commented:
yeah sorry, i don't know why I did up/down... I made those changes and added a default run() method that will give the user a menu of choices to create/remove/clear/help....

philsturgeon commented:
This looks fine now. If you can document it and comment with a link to that pull request in this one then we can merge it.

Commit:
Added better documentation, added comment with link to pull request.

Usage:
Run this task to set add/remove/clear the db sessions table.
Table name will be generated from your config file for your app.
This could be expanded in app/tasks for application specific stuff.

Usage:
php oil r session         = will prompt with menu
php oil r session:create  = create the db table.
php oil r session:remove  = remove the sessions table
php oil r session:clear   = clear the sessions table

dberry37388 commented:
Documentation pull request https://github.com/fuel/docs/pull/310

Commit:
signing it off

Signed-off-by: Daniel Berry <daniel@berrymediagroup.com>

aranw commented:
Just wondering does this need a check for if a database is configured? or perhaps a specific connection? 

Some people may use different databases for sessions. 

Other than that this looks good and I can imagine myself using :) 

+1 for this.

dberry37388 commented:
It currently uses the sessions config file to check the driver to make sure it is "db". If it is not "db", than it tells you its not.  If your actual database is not configured, it will fail with standard error message.

frankdejonge commented on tasks/session.php:
```diff
@@ -0,0 +1,169 @@
+        switch($option)
+        {
+            case "create":
+                return self::create();
```
This should be static::

Commit:
updated functions to static

Updated the functions to static::

Signed-off-by: Daniel Berry <daniel@berrymediagroup.com>

frankdejonge commented on tasks/session.php:
```diff
@@ -0,0 +1,169 @@
+        else
+        {
+            // driver is not set to db, so inform the user.
+            return \Cli::color("Oops, your driver is currently set to ".\Config::get('session.driver').'. Please set your driver type to db before continuing.', 'red');
```
Somebody could still want to do this even if db is not the default driver. A notice would be fine, but somebody should still be able to perform the task.

frankdejonge commented on tasks/session.php:
```diff
@@ -0,0 +1,169 @@
+                php oil r session:create
+                php oil r session:remove
+                php oil r migrate:clear
+                php oil r migrate:help
```
this should be session:clear and session:help

frankdejonge commented on tasks/session.php:
```diff
@@ -0,0 +1,169 @@
+<?
```
The opening tag should include the 'php' part, part of our coding standards.

frankdejonge commented:
First of all, nice job. I think a lot of people will like this. I added some line notes here and there, minor things.

Commit:
allow for created with all driver types

A few minor changes.
Completed the opening <?php tag.

Allowed the user to continue to add the database table regardless of the driver type. This will display a notification to the user, telling them to use the table they will need to set the driver type to "db".

Fixed the "migrate" tags in the help method, changed to "session".


Signed-off-by: Daniel Berry <daniel@berrymediagroup.com>

Commit:
cleaned up code

cleaed the code, added consistent usage of the single and double quotes.


Signed-off-by: Daniel Berry <daniel@berrymediagroup.com>

dberry37388 commented:
does this latest commit look good or is there anything else I need to clean up?

sergiigrebeniuk commented on tasks/session.php:
```diff
@@ -0,0 +1,178 @@
+ * @version    1.1
+ * @author     Fuel Development Team
+ * @license    MIT License
+ * @copyright  2010 - 2011 Fuel Development Team
```
Very minor - it's 2012 now :)

Commit:
updated copyright to 2012

Signed-off-by: Daniel Berry <daniel@berrymediagroup.com>

philsturgeon commented on tasks/session.php:
```diff
@@ -0,0 +1,178 @@
+ * php oil r session:clear   = clear the sessions table
+ */
+
+class Session {
```
We've started dropping these brackets to the line below, as we do it with functions and whatnot.

philsturgeon commented on tasks/session.php:
```diff
@@ -0,0 +1,178 @@
+
+            if ($continue === 'n')
+            {
+                return \Cli::color('Database sessions table was not created.', 'red');
```
You can just throw exceptions and it will do this for you.

dberry37388 replied:
When I throw an exception it gives all the extra in for like ....in APPPATH/tasks/session.php on line 78..... and really do not need that in this instance. And like for example the one above is not an exception, it's just a message to the user.

philsturgeon replied:
Exception shouldnt do that, but I guess it doesnt much matter either way.

dberry37388 replied:
the full error message it gives when outputted is:

``` php
Error - Database sessions table was not created. in APPPATH/tasks/session.php on line 78
```

I would normally use an exception for throwing an actual error or something i could try/catch, but that's not really the case here.  In this case, really just want to show the use a message.

philsturgeon commented on tasks/session.php:
```diff
@@ -0,0 +1,178 @@
+                'created'      => array('constraint' => 10, 'type' => 'int', 'unsigned' => true),
+                'updated'      => array('constraint' => 10, 'type' => 'int', 'unsigned' => true),
+                'payload'      => array('type' => 'longtext'),
+            ), array('session_id'), false, 'InnoDB', 'utf8');
```
Is assuming InnoDB a good thing?

Also, utf8 is a config option, we dont need to specify this.

dberry37388 replied:
if I am just selecting data and making a few updates here and there I generally go with MyISAM.... But when it comes to heavy selects and update requests constantly battering the db, InnoDB is the way to go. You have to take into account the table vs. row locking , crash response, etc...

dberry37388 commented:
I had tried to leave this as it was in the session config docs, so that the generated SQL would be the same as what was provided in the example just for consistency.

philsturgeon commented:
The session config docs are an example, so setting this explicitly is not required.

Commit:
updated opening class tag & charset to match config

moved the bracket on the opening class tag down to a newline... That to me just feels and looks weird though...
also changed the hard-coded 'utf8' charset in the create sql to \Config::get() to pull from the user's config.

Signed-off-by: Daniel Berry <daniel@berrymediagroup.com>
